<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>PEACHTree</title>
	<link rel="stylesheet" href="../css/styles.css">
	<link rel="icon" href="../img/icon.png">
    <script defer src="../cpp/peachtree/peachtree.js"></script>

  	<script language='JavaScript' type='text/JavaScript' src="https://code.jquery.com/jquery-1.12.4.js"></script>
  	<script language='JavaScript' type='text/JavaScript' src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>


	<script language='JavaScript' type='text/JavaScript' src="../js/WW_controller.js"></script>
	
		
  </head>
  <body>
  </body>
  <script>
  
		FADE_IN_TIME = 100;
		WEB_WORKER_WASM = null;
  
		function init() {



			//addLoader($("#main"), true);

			register_WebWorker(function(){

				callWasmFunction("init", [], function(){

					// Render options
					callWasmFunction("getOptions", [], function(options){
					
					
						// Get the unique list of sections
						var sections = [];
						for (var i = 0; i < options.length; i ++){
							let opt = options[i];
							if (opt.hide) continue;
							let s = opt.section;
							if (!sections.includes(s)) sections.push(s);

						}
						
						
	
						
						// Render options
						//console.log(options);
						for (let i = 0; i < sections.length; i++){
						
							let s = sections[i];
						
							// Create div for each section
							let id = "sect_" + s;
							let optHTML = `<div class="sectionDiv" id="` + id + `">
												<div class="sectionHeader">`
													+ s + 
												`</div>
											</div>`;
							$("#displayOptions").append(optHTML);
						
						
							// Get the options of this section
							for (let o = 0; o < options.length; o ++){
								let opt = options[o];
								if (opt.hide) continue;
								if (opt.section == s){
								
									var type = opt.type;
									var val = opt.value + "";
									if (val != "" && val != null) val = " <i>Default: " + val + "</i>";
									else val = "";
									console.log(s, opt);
									$("#" + id).append(`<div class="optionDiv"> <span class="parameterName"><span class="parameterType">` + type + `</span> ` + opt.name + `:</span> ` + opt.longTitle + val + `<br>`);
								
								
								}
									
							}

						}
					
					});

					
				});


			}, "../cpp/peachtree/WW_model.js")

			
	  
	  }
  </script>
  
  <body onload="init()">
  
  
  
	<div id="main">
	
		<div id="mainHeader"  style="text-align:center">
			<h1> PEACH Tree </h1>
			<h4> <b>P</b>lotting <b>E</b>pidemiological and <b>A</b>lignment <b>CH</b>aracters onto phylogenetic <b>Tree</b>s </h4>
		</div>


		
		
		<div id="about_div">
		
		
			<h1>PEACH Tree Documentation</h1>
			
			
			
					
				
				<ul>
				
					<li>Upload a multiple sequence alignment of biological sequences</li>
					
					<li>Upload a phylogenetic tree, or press 
						<span style="margin:auto;" id="buildTreeBtn" class="button small" title="Build a neighbour-joining tree">Build tree</span>
						and we will build the neighbour-joining tree from the alignment </li>
						
	
					<li> Only the segregating sites are shown, so that you are not overwhelmed with information </li>
					
					
					<li> Samples can be filtered and the the segregating sites are recalculated  </li>
					
					
					<li> Sample dates and symptom onset dates can be displayed on the transmission tree  </li>
				
				</ul>


				<img src="../img/Fig1.png" height="400px">


				<div id="softwareDescDiv">
					<b>PEACH Tree</b> is a high-performance web tool for visualising multiple sequence alignments, and is specifically designed for epidemiological domains, such as SARS-CoV-2 pathogen surveillance. 
					By default, only segregating alignment sites (i.e. sites which vary among the samples) are displayed, making the program suitable for viewing full genomic data.
					A phylogenetic tree can be generated from the alignment or uploaded separately.
					Epidemiological annotations can be uploaded (such as symptom onset date) and viewed alongside the rest of the data.
					PEACH Tree is an effort to combine genomic and epidemiological modelling into a unified framework.
					<br><br>
					
		
				</div>
				
				
		
			
			<!-- Subsetting tree -->
			<div class="section">
			
			
				<div class="heading">
					Filtering samples
				
				</div>
				
				
				<div class="desc">
					

					To filter samples out of the display, samples must be selected first.
					These can be selected (or deselected) by clicking on their accession names or by searching for their accessions in the searchbar.
					Use shift+click to select multiple consecutive samples.
					When samples are filtered out of the display, the segregating sites are recalculated.
					To control sample filtering: <br><br>
					
					<div class="optionBtnDiv">
						<span id="btn_focus" class="button small" title="Focus on taxa selection (click on a label below to select)">Focus</span>: 
								filter out all unselected samples. 
					</div>
					
					<div class="optionBtnDiv">
						<span id="btn_clade" class="button small" title="Focus on selected subtree/clade (only applicable if a tree is uploaded)">Subtree</span>: 
								filter out all samples which are not monophyletic (i.e., form a <a href="https://en.wikipedia.org/wiki/Clade" target="_blank">clade</a>) with the selected samples. Requires a tree. 
					</div>
					
					<div class="optionBtnDiv">
						<span id="btn_clear" class="button small" title="Clear selection">Clear</span>: 
								clear the selection and display all samples. 
					</div>

					
					
					
					
					
				
				</div>
				

			
			</div>
			
			
			
			
			<!-- Displaying epi info -->
			<div class="section">
			
			
				<div class="heading">
					Displaying epidemiological information
				
				</div>
				
				
				<div class="desc">
					

					Epidemiological information can be displayed on the tree or in place of sequence accession names.
					Metadata must be uploaded first - ensuring that one of the columns is named "accession".
					If there is a tree, open the "Epidemiology" tab and set the "Sample date" to the appropriate date variable, and a calendar will be added to the tree. 
					This method requires at least 2 unique dates.  
					Afterwards, the symptom onset dates of cases can be displayed on the transmission tree (plus and minus a time period to reflect infectious period).
					
					
					
					<br><br>
					

					
					
					
					
				
				</div>
				

			
			</div>
			
			
			
			
			<!-- Display all options here -->
			<div class="section">
			
			
				<div class="heading">
					Display Options
				
				</div>
				
				<div id="displayOptions" class="desc">
				
					Display options can be adjusted using the tabs at the bottom of the main page. 
					They can also be set from the command line using the variable names below.
				
					
				
				</div>

			
			</div>
			
			
			<!-- Command line -->
			<div class="section">
			
			
				<div class="heading">
					Download
				
				</div>
				
				
				<div class="desc">
					

					
					PEACH Tree can be run in the command line to generate svgs from input files and optional display options. 
					To do this, first download PEACH Tree from GitHub
					
					
					<div class="code">
						git clone https://github.com/phylopeachtree/phylopeachtree.github.io
					</div>
					
					Then run the C++ executable using
					
					
					<div class="code">
						./phylopeachtree.github.io/cpp/peachtree/peachtree -h
					</div>
					
					with -h enabled to view command line options.
					
					
					
				
				</div>
				

			
			</div>



		</div>






	
	</div>


	<div id="slideIns">


	</div>
	
  
	
  
  </body>
  
</html>